### PR TITLE
[compiler] Support for context variable loop iterators

### PR DESCRIPTION
**Pull request by mvitousek**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #31712
* #31710
* __->__ #31709

Summary:
Fixing a compiler todo

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts [hunk 1]
```diff
@@ -1378,20 +1364,33 @@ function codegenForInit(
...
-        {
-          reason: 'Expected a variable declaration',
-          loc: init.loc,
-          description: `Got ${instr.type}`,
-          suggestions: null,
-        },
-      );
-      if (instr.kind === 'let') {
-        kind = 'let';
+      let top: undefined | t.VariableDeclarator = undefined;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts:1367]: where is `top` used other than being assigned to?

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts [hunk 2]
```diff
@@ -1378,20 +1364,33 @@ function codegenForInit(
...
+      if (
+        instr.type === 'ExpressionStatement' &&
+        instr.expression.type === 'AssignmentExpression' &&
+        instr.expression.operator === '=' &&
+        instr.expression.left.type === 'Identifier' &&
+        (top = declarators.at(-1))?.id.type === 'Identifier' &&
+        top?.id.name === instr.expression.left.name &&
+        top?.init == null
+      ) {
+        top.init = instr.expression.right;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts:1377]: isn't this mutating the ast?

**Review by josephsavona — APPROVED** [review]

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts (hunk 2 above)
Comment (mvitousek) [reply compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts:1377]: Yes, I think that's ok at this stage? The node being mutated hasn't been attached to anything else yet. But can definitely just delete the declarator being mutated and re-create it instead

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts (hunk 1 above)
Comment (mvitousek) [reply compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts:1367]: Body of the if and the other if conditions

Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts:1367]: Yeah but it looks like we're reading in the if condition, modifying in the if consequent, but then not using the value again? I guess a better question is why do we need to assign to `top.init` in the body? I'm probably forgetting some surrounding context here.

Comment (mvitousek) [reply compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts:1367]: Oh it's an alias to the last declarator that was pushed to the list of variable declarators. Basically if we have a context iterator, the HIR is a DeclareContext followed by a StoreContext; when we lower that to JS it becomes a variable declarator without an init followed by an assingment to that variable, and here `top` is a reference to the uninitialized declarator, and we're setting the declarator's init to tbe the RHS of the assignment to knit everything back up.

**Pull request merged**